text/0088-tvm-update.md

```diff
@@ -0,0 +1,142 @@
+
+**12th** element of **c7** tuple contains fee debited in storage phase.
+
+**13th** elements of **c7** contains of tuple with two elements: `last_mc_blocks` and `prev_key_block`.
+- `last_mc_blocks` is a tuple of up to 16 elements which contains previous most recent masterchain blocks info (ordered by seqno). Each block is represented as tuple of 5 integers: `(workchain, shard, seqno, root_hash, file_hash)`
+- `prev_key_block` is a tuple of 5 integers `(workchain, shard, seqno, root_hash, file_hash)` corresponding to most recent key block.
+
+## TVM update
+### New stack type
+This proposal suggests to introduce new type in TVM `Hasher`: it is opaque type used to consequently hash data chunks. This type behave the same way as others and can be considered immutable: `HASHAPP*` ops (described below) "consume" old hasher and return updated one (that way if old hasher was copied, it's copies are completely independent). This object is not serializable and thus can not be returned from TVM (for instance as result of get-method).
+### New opcodes
+#### c7 and block ops
+`MYCODE` - `0xF82A` - ` - c` - return cell with currently executed code. Equivalent to `10 GETPARAM`.
+
+`INCOMINGVALUE` - `0xF82B` - ` - i` - return integer with TON amount in incoming message (or zero if not applicable). Equivalent to `11 GETPARAM`.
+
+`STORAGEFEES` - `0xF82C` - ` - i` - return integer value of fee in nanoTONs debited in storage phase. Equivalent to `12 GETPARAM`.
+
+`PREVBLOCKSINFOTUPLE` - `0xF82D` - ` - t` - return tuple with previous blocks info. Equivalent to `13 GETPARAM`.
+
+`PREVMCBLOCKS`- `0xF83400` - ` - t` - return tuple with previous masterchain blocks info. Equivalent to `13 GETPARAM; FIRST`
+
+`PREVKEYBLOCK`- `0xF83401` - ` - t` - return tuple with previous keyblock id. Equivalent to `13 GETPARAM; SECOND`
+
+#### Hash primitives
+`HASHSTART_SHA256` - `0xF90300` - ` - h` - create `sha256` hasher object.
+
+`HASHSTART_SHA512` - `0xF90301` - ` - h` - create `sha512` hasher object.
+
+`HASHSTART_BLAKE2B` - `0xF90302` - ` - h` - create `blake2b` hasher object.
+
+`HASHSTART_KECCAK256` - `0xF90303` - ` - h` - create `keccak256` hasher object.
+
+`HASHSTART_KECCAK512` - `0xF90304` - ` - h` - create `keccak512` hasher object.
+
+
+`HASHEND` - `0xF904` - `h - ... ` - calculate hash from hasher and put in on the stack. If size of the hash does not exceed 256 bits, it is returned as a 256-bit unsigned integer (e.g. sha256). Otherwise it is returned as a tuple of 256-bit integers (e.g. sha512 - tuple of two integers). If the bit length of data in hasher is not divisible by eight, throws a cell underflow exception.
+
+`HASHENDST` - `0xF905` - `b h - b' ` - calculate hash from hasher and store it into the slice.
+
+
+`HASHINFO` - `0xF906` - `h - i` - return hash type of hasher (`sha256` - `0`, `sha512` - `1`, `blake2b` - `2`, `keccak256` - `3`, `keccak512` - `4`).
```

Comment:
I think that hardcoding magic numbers is always a bad practice. There are many algorithms with different names and for sure in the future the list of them will be replenished. If at this stage it is assumed that the ID for the hasher will be chosen purely for "historical reasons", then for many it will cause misunderstanding. I suggest using the same function instead of ordinal numbers as for determining the get-methods ID in smart contracts